src/main/java/com/googleintern/wfm/ruleengine/action/GeneralRulesForWorkgroupId.java

```diff
@@ -0,0 +1,76 @@
+package src.main.java.com.googleintern.wfm.ruleengine.action;
+
+import com.google.common.collect.*;
+import src.main.java.com.googleintern.wfm.ruleengine.model.FilterModel;
+import src.main.java.com.googleintern.wfm.ruleengine.model.PoolAssignmentModel;
+import src.main.java.com.googleintern.wfm.ruleengine.model.RuleModel;
+import src.main.java.com.googleintern.wfm.ruleengine.model.UserPoolAssignmentModel;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * GeneralRulesForWorkgroupId class is used to create rules that can apply to all users from the
+ * same work group.
+ */
+public class GeneralRulesForWorkgroupId {
+  /** Generate rules that can apply to all users from the same workgroup Id. */
+  public static ImmutableSet<RuleModel> generalRuleByWorkgroupId(
+      ImmutableListMultimap<Long, UserPoolAssignmentModel> userPoolAssignmentsByWorkGroupIdBuilder,
+      Long workgroupId) {
+    ImmutableList<UserPoolAssignmentModel> userPoolAssignmentsFromSameWorkGroupId =
+        userPoolAssignmentsByWorkGroupIdBuilder.get(workgroupId);
+    ImmutableSet<PoolAssignmentModel> permissionIntersections =
+        findPermissionIntersection(userPoolAssignmentsFromSameWorkGroupId);
+    ImmutableSetMultimap<Long, Long> permissionGroup =
+        groupPermissionByCasePoolId(permissionIntersections);
+    Long workforceId = userPoolAssignmentsFromSameWorkGroupId.get(0).workforceId();
+    return createGeneralRuleForWorkgroupId(permissionGroup, workforceId, workgroupId);
+  }
+
+  private static ImmutableSet<PoolAssignmentModel> findPermissionIntersection(
+      List<UserPoolAssignmentModel> userPoolAssignmentsFromSameWorkGroupId) {
+    ImmutableSet<PoolAssignmentModel> permissionIntersections =
+        userPoolAssignmentsFromSameWorkGroupId.get(0).poolAssignments();
+
+    for (UserPoolAssignmentModel user : userPoolAssignmentsFromSameWorkGroupId) {
+      permissionIntersections =
+          ImmutableSet.copyOf(Sets.intersection(permissionIntersections, user.poolAssignments()));
+      if (permissionIntersections.size() == 0) {
+        return ImmutableSet.of();
+      }
+    }
```

Comment:
OPTIONAL
This logic can be done with Java8 Streams and reduce:
```
return ImmutableSet.copyOf(userPoolAssignmentsFromSameWorkgroupId
   .stream()
   .reduce(userPoolAssigmentsFromSameWorkgroupId.get(0).poolAssignments(),
 (intersections, user) -> Sets.intersection(intersections, user.poolAssignments())));
```

Reply:
I tried to use the lambda expression. It  gives me an error for `intersections` variable in the sets.intersection() function. 
intersections: 
Required type: Set
Provided: UserPoolAssignmentModel
I searched online and tried to add combiner to reduce().
`ImmutableSet.copyOf(
        userPoolAssignmentsFromSameWorkGroupId.stream()
            .reduce(
                userPoolAssignmentsFromSameWorkGroupId.get(0).poolAssignments(),
                (intersections, user) -> Sets.intersection(intersections, user.poolAssignments()),
                    Sets::intersection));`
It gives me an error about recognizing object E for Set<E>.

Reply:
This one is my fault for getting the types wrong. Stream.reduce combines a list of type T into just a single T. I misinterpreted the objects here, and wrote code for List<Set<PoolAssignmentModel>> instead of List<UserPoolAssignmentModel>. To fix this, we will need to add a map from each UserPoolAssignmentModel to its Set<PoolAssignmentModel>
```
return ImmutableSet.copyOf(userPoolAssignmentsFromSameWorkgroupId
   .stream()
   .map( userPoolAssignment -> user.poolAssignments())
   .reduce(userPoolAssigmentsFromSameWorkgroupId.get(0).poolAssignments(),
 (intersections, user) -> Sets.intersection(intersections, user.poolAssignments())));
```